executable-spec/src/Cardano/Ledger/Spec/STS/Update/Tallysip.hs

```diff
@@ -0,0 +1,222 @@
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Cardano.Ledger.Spec.STS.Update.Tallysip where
+
+import           Data.Map.Strict (Map, (!))
+import qualified Data.Map.Strict as Map
+import           Data.Set as Set (Set)
+import qualified Data.Set as Set
+import           GHC.Generics (Generic)
+import           Data.AbstractSize (HasTypeReps)
+
+import           Control.State.Transition (Embed, Environment, IRC (IRC),
+                     PredicateFailure, STS, Signal, State, TRC (TRC),
+                     initialRules, judgmentContext,
+                     trans, transitionRules, wrapFailed, (?!))
+import           Ledger.Core (Slot, addSlot, (⨃), (∈), (∉), dom)
+import qualified Ledger.Core as Core
+import           Cardano.Crypto.Hash (HashAlgorithm)
+
+import qualified Cardano.Ledger.Spec.STS.Update.Data as Data
+
+-- | STS for tallying the votes of a single SIP
+data TALLYSIP hashAlgo
+
+data Env hashAlgo
+ = Env { currentSlot :: !Slot
+       , sipdb :: !(Map (Data.SIPHash hashAlgo) (Data.SIP hashAlgo))
+       , ballots :: !(Map (Data.SIPHash hashAlgo) (Map Core.VKey Data.Confidence))
+       , vThreshold :: !Data.VThreshold
+       , stakeDist :: !(Map Core.VKey Data.Stake)
+       }
+       deriving (Eq, Show)
+
+data St hashAlgo
+  = St { vresips :: !(Map (Data.SIPHash hashAlgo) Data.VotingResult)
+         -- ^ Records the current voting result for each SIP
+       , asips :: !(Map (Data.SIPHash hashAlgo) Slot)
+      -- ^ Active SIP's. The slot in the range (of the map) determines when the
+      -- voting period will end.
+       , apprvsips :: !(Set (Data.SIPHash hashAlgo))
+         -- ^ Set of approved SIPs
+       }
+       deriving (Eq, Show, Generic)
+
+
+instance STS (TALLYSIP hashAlgo) where
+
+  type Environment (TALLYSIP hashAlgo) = (Env hashAlgo)
+
+  type State (TALLYSIP hashAlgo) = (St hashAlgo)
+
+  type Signal (TALLYSIP hashAlgo) = (Data.SIPHash hashAlgo)
+
+  data PredicateFailure (TALLYSIP hashAlgo)
+    =
+      TallySIPFailure (Data.SIPHash hashAlgo)
+    | InvalidSIPHash (Data.SIPHash hashAlgo)
+    | SIPAlreadyApproved (Data.SIPHash hashAlgo)
+    | VoteFromInvalidStakeholder (Data.SIPHash hashAlgo)
+    deriving (Eq, Show)
+
+  initialRules = [
+      do
+        IRC Env { } <- judgmentContext
+        pure $! St { vresips = Map.empty
+                   , asips = Map.empty
+                   , apprvsips = Set.empty
+                   }
+      ]
+
+  transitionRules = [
+    do
+      TRC ( Env { currentSlot
+                , sipdb
+                , ballots
+                , vThreshold
+                , stakeDist
+                }
+          , St  { vresips
+                , asips
+                , apprvsips
+                }
+          , sipHash
+          ) <- judgmentContext
+
+      sipHash ∈ dom sipdb ?! InvalidSIPHash sipHash
+
+      sipHash ∉ apprvsips ?! SIPAlreadyApproved sipHash
+
+      -- all ballots should come from valid stakeholders
+      case Map.lookup sipHash ballots of
```

Comment:
Why do we check this here, and not when adding a vote to `ballots`?

Reply:
We do check it when we add a ballot (based on `participants`). The `stakeDist` is indeed generated from `participants`, so there is no need for a check here. I will remove it.

Reply:
Done OK